data/src/main/java/org/apache/iceberg/data/PartitionStatsHandler.java

```diff
@@ -135,20 +143,113 @@ public static PartitionStatisticsFile computeAndWriteStatsFile(Table table) thro
    */
   public static PartitionStatisticsFile computeAndWriteStatsFile(Table table, long snapshotId)
       throws IOException {
+    return computeAndWrite(table, snapshotId, true /* recompute */);
+  }
+
+  /**
+   * Incrementally computes the stats after the snapshot that has partition stats file till the
+   * given snapshot and writes the combined result into a {@link PartitionStatisticsFile} after
+   * merging the stats.
+   *
+   * @param table The {@link Table} for which the partition statistics is computed.
+   * @param snapshotId snapshot for which partition statistics are computed.
+   * @return {@link PartitionStatisticsFile} for the given snapshot, or null if no statistics are
+   *     present.
+   */
+  public static PartitionStatisticsFile computeAndWriteStatsFileIncremental(
+      Table table, long snapshotId) throws IOException {
+    return computeAndWrite(table, snapshotId, false /* recompute */);
+  }
+
+  private static PartitionStatisticsFile computeAndWrite(
+      Table table, long snapshotId, boolean recompute) throws IOException {
+    Preconditions.checkArgument(table != null, "Table cannot be null");
     Snapshot snapshot = table.snapshot(snapshotId);
     Preconditions.checkArgument(snapshot != null, "Snapshot not found: %s", snapshotId);
 
-    Collection<PartitionStats> stats = PartitionStatsUtil.computeStats(table, snapshot);
+    StructType partitionType = Partitioning.partitionType(table);
+
+    Collection<PartitionStats> stats;
+    if (recompute) {
+      stats = PartitionStatsUtil.computeStats(table, snapshot);
+    } else {
+      stats = incrementalComputeAndMerge(table, snapshot, partitionType);
+    }
+
     if (stats.isEmpty()) {
       return null;
     }
 
-    StructType partitionType = Partitioning.partitionType(table);
     List<PartitionStats> sortedStats = PartitionStatsUtil.sortStats(stats, partitionType);
     return writePartitionStatsFile(
         table, snapshot.snapshotId(), schema(partitionType), sortedStats);
   }
 
+  private static Collection<PartitionStats> incrementalComputeAndMerge(
+      Table table, Snapshot snapshot, StructType partitionType) throws IOException {
+    PartitionStatisticsFile statisticsFile = latestStatsFile(table, snapshot.snapshotId());
+    if (statisticsFile == null) {
+      throw new RuntimeException(
+          "Previous stats not found for incremental compute. Try full compute");
+    }
+
+    PartitionMap<PartitionStats> statsMap = PartitionMap.create(table.specs());
+    // read previous stats, note that partition field will be read as GenericRecord
+    try (CloseableIterable<PartitionStats> oldStats =
+        readPartitionStatsFile(schema(partitionType), Files.localInput(statisticsFile.path()))) {
```

Comment:
Just brainstorming: If the referenced stat file doesn't exist we throw some data reading specific exception here (haven't found a test to see what exception, but probably I wasn't searching hard enough). What I have in mind is that we could wrap it into an exception that is relevant for the incremental stats computing. Like `IncrementalComputeStatsFailedException` or such. Then we could include a member in this specific exception that holds the snapshot ID of `statisticsFile`. On the caller side then the caller could decide to fix this broken state of the table/stats and invoke compute stats for that particular snapshot.
LMK WDYT!